sql: use global defaults for internal session data rather than zero-values

rafiss opened the pull request:
fixes https://github.com/cockroachdb/cockroach/issues/70888

Release note: None

rafiss committed:
sql: use global defaults for internal session data rather than zero-values

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/101486)
<!-- Reviewable:end -->

rafiss committed:
*: change newInternalPlanner so it can use non-zero session data

Release note: None

rafiss committed:
sql: rename NewFakeSessionData to NewInternalSessionData

Release note: None

rafiss committed:
changefeedccl: use default sessiondata instead zero sessiondata

Release note: None

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101486)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benbardin, @mgartner, @rafiss, and @shermanCRL)

___
*[`pkg/sql/internal.go` line 56 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/101486#-NSwiTIV2NR29ab5O1uT:-NSwiTIV2NR29ab5O1uU:b-zf9ce4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d02a8dde44bcbd6106ead286eb609986474a356a/pkg/sql/internal.go#L56)):*
> ```Go
> // steps of background jobs and schema changes. Each session variable is
> // initialized using the correct default value.
> func NewInternalSessionData(
> ```

What makes this "internal"? Why is it instantiated differently from the user SessionData?


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101486)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benbardin, @ecwall, @mgartner, and @shermanCRL)

___
*[`pkg/sql/internal.go` line 56 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/101486#-NSwiTIV2NR29ab5O1uT:-NSxWg8G0JS8AeCg9TSM:b-crkdgo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d02a8dde44bcbd6106ead286eb609986474a356a/pkg/sql/internal.go#L56)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

What makes this "internal"? Why is it instantiated differently from the user SessionData?
</blockquote></details>

there are two things that make it internal:
- the `opName` is passed in to indicate which internal operation is being executed
- a real user session has `SessionArgs` which come from the connection string parameters, and also includes defaults that are configured by `ALTER ROLE username SET varname = ...`. we don't have this here.


<!-- Sent from Reviewable.io -->

rafiss committed:
sql: add a TODO for sql-queries to investigate a test

After changing the sessiondata to stop using zero-value defaults, a test
started failing. It works when using a zero sessiondata, which deserves
more investigation.

Release note: None

rafiss committed:
sql: use NewInternalSessionData for InternalDB

Prevent zero-values from being used for session vars when creating an InternalDB.

Release note: None

rafiss committed:
schemachanger: fix upgrade jobs to work with declarative schema changer

Now that the internal executor is using a non-zero session data, that
means it uses the declarative schema changer by default. The upgrade
manager had a few assumptions in it that were specific to the old schema
changer:

- The PGAttributeNum field should only be set if it differs from the
  column descriptor ID.
- Check the DeclarativeSchemaChangerState when waiting for migrations to
  finish.

Release note: None

ecwall reviewed (commented):
> Reviewed 2 of 10 files at r1, 2 of 53 files at r5, 25 of 60 files at r6, 18 of 22 files at r7, 4 of 6 files at r8, 1 of 1 files at r9, 2 of 3 files at r10, 1 of 52 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101486)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benbardin, @mgartner, @rafiss, and @shermanCRL)

___
*[`pkg/sql/internal.go` line 57 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/101486#-NTEYjCA4WWaTOhhJmV3:-NTEYjCA4WWaTOhhJmV4:bi6ntyg) ([raw file](https://github.com/cockroachdb/cockroach/blob/236414b49a40803e2866320f99093c8d547285b5/pkg/sql/internal.go#L57)):*
> ```Go
> // steps of background jobs and schema changes. Each session variable is
> // initialized using the correct default value.
> func NewInternalSessionData(
> ```

It looks like a few other params might be useful here like the setting database and the code I mentioned here https://reviewable.io/reviews/cockroachdb/cockroach/101486#-NTEXaA_8kM71fqLJX-O (feel free to do this in a separate ticket(s) since this PR already cleans up a lot).
___
*[`pkg/sql/planner.go` line 374 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/101486#-NTEXaA_8kM71fqLJX-O:-NTEXaA_8kM71fqLJX-P:biamhyx) ([raw file](https://github.com/cockroachdb/cockroach/blob/236414b49a40803e2866320f99093c8d547285b5/pkg/sql/planner.go#L374)):*
> ```Go
> 	sd.SessionData.UserProto = user.EncodeProto()
> 	sd.SessionData.Internal = true
> 	sd.SearchPath = sessiondata.DefaultSearchPathForUser(user)
> ```

`newInternalPlanner` looks like a constructor function so I would not expect it to also modify the input `sd`.

Can you extract the code modifying `sd` and call it before calling `newInternalPlanner` (maybe in a separate ticket)?


<!-- Sent from Reviewable.io -->

benbardin commented on pkg/ccl/streamingccl/streamproducer/replication_manager_test.go:
```diff
@@ -45,16 +46,20 @@ func TestReplicationManagerRequiresReplicationPrivilege(t *testing.T) {
 	var sessionSerialized []byte
 	tDB.QueryRow(t, "SELECT crdb_internal.serialize_session()").Scan(&sessionSerialized)
-	require.NoError(t, protoutil.Unmarshal(sessionSerialized, &sessionData))
+	require.NoError(t, protoutil.Unmarshal(sessionSerialized, &m))
```
Will this work on a mixed-version cluster?

rafiss replied:
yes, this will be fine, since adding new fields to a proto is backwards compatibile

chengxiong-ruan reviewed (approved):
very nice!

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101486)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benbardin, @ecwall, @mgartner, and @shermanCRL)

___
*[`pkg/sql/planner.go` line 374 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/101486#-NTEXaA_8kM71fqLJX-O:-NUC5TKFElxzHbZSN-oR:b-nwt0e2) ([raw file](https://github.com/cockroachdb/cockroach/blob/236414b49a40803e2866320f99093c8d547285b5/pkg/sql/planner.go#L374)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

`newInternalPlanner` looks like a constructor function so I would not expect it to also modify the input `sd`.

Can you extract the code modifying `sd` and call it before calling `newInternalPlanner` (maybe in a separate ticket)?
</blockquote></details>

i added a call to `clone` so the input isn't modified


<!-- Sent from Reviewable.io -->

rafiss commented:
tftr!

bors r=chengxiong-ruan
